### PR TITLE
Add new docker container resource

### PR DESCRIPTION
**Pull request by djui**
This allows to await for a `healthy` docker container's healthcheck status.

The motivation for this comes from a discussion about the new `HEALTHCHECK` feature in Docker and its lack of support in Docker-Compose: Developers write Makefiles(or other scripts) that start a service declared by Docker-Compose, wait until its health-check passes, and then start the next service. Now they can use `await` outside Docker-Compose in their scripts.

Examples:

```
$ docker ps
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS                    PORTS                    NAMES
fe13f7f4a1b3        helloworld               "/helloworld.bin"        3 hours ago         Up 34 minutes             0.0.0.0:32777->80/tcp    foo_fighter
fe13f7f4a1b4        helloworld:healthcheck   "/helloworld.bin"        3 hours ago         Up 34 minutes (healthy)   0.0.0.0:32777->80/tcp    big_hamilton
```

```
$ ./await -vv -t 30s docker:///fe13f7f4a1b4
2016/12/03 23:26:35 Awaiting resource: docker://localhost/foo
2016/12/03 23:26:35 Resource unavailable: container not running
:
2016/12/03 23:26:40 Resource unavailable: health status: starting
:
2016/12/03 23:26:45 All resources available
```

```
$ ./await -vv -t 30s docker:///var/run/docker.sock/fe13f7f4a1b4
2016/12/03 23:26:35 Awaiting resource: docker://localhost/foo
2016/12/03 23:26:35 Resource unavailable: container not running
:
2016/12/03 23:26:40 Resource unavailable: health status: starting
:
2016/12/03 23:26:45 All resources available
```

```
$ ./await -vv -t 30s docker://#name=big_hamilton
2016/12/03 23:26:35 Awaiting resource: docker://localhost/foo
2016/12/03 23:26:35 Resource unavailable: container not running
:
2016/12/03 23:26:40 Resource unavailable: health status: starting
:
2016/12/03 23:26:45 All resources available
```

```
$ ./await -vv -t 30s docker://#name=foo_fighter
2016/12/04 01:42:55 Awaiting resource: docker://localhost/foo
2016/12/04 01:42:55 Resource unavailable: container without health check
:
2016/12/04 01:43:25 Timeout exceeded
```

```
$ ./await -vv -t 30s docker://#name=not_exists
2016/12/03 23:10:45 Awaiting resource: docker://#name=not_exists
2016/12/03 23:10:45 Error: failed to await resource: container not found
:
2016/12/03 23:11:15 Timeout exceeded
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -150,6 +151,30 @@ classified as available as soon as the database was found.
...
+
+**Availability**: Available when a given Docker container is running, not paused
+and not restarting, and has a health check status set to `healthy`. Unavailable
+otherwise.
+
+**URL syntax**: `docker://[<host>[:<port>]][/<container_id>][#<fragment>]`
+
+The `<host>` (and optional `<port>`) specifies the Docker daemon host. It can be
+either a unix socket file path, TCP URL host, or absent. If absent, the
+`DOCKER_HOST` environment variable's value will be used other
```
Comment (djui) [inline README.md]: Should have been _otherwise_

File: README.md [hunk 2]
```diff
@@ -150,6 +151,30 @@ classified as available as soon as the database was found.
...
+`unix:///var/run/docker.sock`.
+
+The `<container_id>` can be optional but then requires the `name` or `image`
```
Comment (djui) [inline README.md]: yes.

**Comment by djui** [comment]
@fcristovao Fixed the grammar issues and added some use cases.

File: README.md [hunk 3]
```diff
@@ -44,6 +51,63 @@ or
...
+    :
+```
+
+Outside Docker-Compose using `HEALTHCHECK`:
+
+```makefile
+test-it:
+	docker-compose up -d
+	await -v \
+	    docker://?name=project_github_mock_1 \
```
Comment (fcristovao) [inline README.md]: Weird indentation here

File: README.md [hunk 4]
```diff
@@ -161,11 +225,11 @@ otherwise.
 
 The `<host>` (and optional `<port>`) specifies the Docker daemon host. It can be
 either a unix socket file path, TCP URL host, or absent. If absent, the
-`DOCKER_HOST` environment variable's value will be used other
+`DOCKER_HOST` environment variable's value will be used, otherwise
 `unix:///var/run/docker.sock`.
 
-The `<container_id>` can be optional but then requires the `name` or `image`
-key-value pair in the `<fragment>`.
+Either `<container_id>` or `<fragment>` with `name` or `image` key-value pair in
```
Comment (fcristovao) [inline README.md]: > Either ... or ... 

with no verb

File: README.md (hunk 3 above)
Comment (djui) [inline README.md]: Continuation lines should always be indented.

Comment (fcristovao) [inline README.md]: I agree, but the problem shows a few lines below.
I'll recomment there...

File: README.md (hunk 4 above)
Comment (djui) [inline README.md]: done.

File: README.md [hunk 5]
```diff
@@ -44,6 +51,63 @@ or
...
+
+Outside Docker-Compose using `HEALTHCHECK`:
+
+```makefile
+test-it:
+	docker-compose up -d
+	await -v \
+	    docker://?name=project_github_mock_1 \
+	    docker://?name=project_sendgrid_1 \
+        docker://?name=project_myservice_db_1 \
```
Comment (fcristovao) [inline README.md]: Compare this line's indentation with above.

Comment (djui) [inline README.md]: Tip: add the comment to the line of problem. Just looking at the comment section does not show enough context.

Comment (djui) [inline README.md]: ok, fixed.

File: README.md [hunk 6]
```diff
@@ -4,9 +4,9 @@
 
 Await availability of resources.
 
-This can be useful in the context of
-[Docker Compose](https://docs.docker.com/compose/) where a service needs to wait
-for other dependent services.
+This can be useful in the context of [Docker Compose](https://docs.docker.com/compose/)
+where a service, just as an integration test driver, needs to wait for other
+dependent services.
```
Comment (djui) [inline README.md]: dfdsfdsfs

Comment (djui) [inline README.md]: ffsfdsf

File: README.md [hunk 7]
```diff
@@ -44,6 +51,63 @@ or
...
+```
+
+Outside Docker-Compose using `HEALTHCHECK`:
+
+```makefile
+test-it:
+	docker-compose up -d
+	await -v \
+	    docker://?name=project_github_mock_1 \
+	docker://?name=project_sendgrid_1 \
```
Comment (fcristovao) [inline README.md]: Weird indentation here

**Comment by lox** [comment]
Would be great to see this land, any progress?

**Comment by djui** [comment]
@lox We currently don't have that use case so I don't want to merge a functionality that might not match real users use case, therefore stalled the progress on this one.

Does the current implementation match your use case? Is something missing or off?

**Comment by lox** [comment]
I'm slowly starting to use container health checks, being able to use them in await would be great. It seems like docker-compose will eventually do this though, so can understand holding off.

**Comment by Joshfindit** [comment]
It seems `docker-compose` is steering away from waiting on health checks and as per https://docs.docker.com/compose/startup-order/ has now depreciated `condition:`.

(They seem to be instead pushing for devs to write 'resilient code' that assumes the db may not be accessible, steering people towards solutions that scale nicely in swarms)

@djui What effect does that have on this PR?

**Comment by Joshfindit** [comment]
As for my own use-case:

I'm just looking for solutions that let me store confidential passwords in environment variables, and it looks like this would be a solution.

**Comment by djui** [comment]
This PR is quite outdated and I would need to recover some context about it to see if it still makes sense. Thus I rather want to see if we address your concrete issue https://github.com/betalo-sweden/await/issues/34 and hopefully that'd sufficient.

**Pull request closed without merging**